joeynmt/decoders.py

```diff
@@ -424,3 +426,102 @@ def _init_hidden(self, encoder_final: Tensor = None) \
+    Decoder layers are masked so that an attention head cannot see the future.
+    """
+
+    def __init__(self, num_layers=4, num_heads=8,
+                 hidden_size=512, ff_size=2048, dropout=0.1,
+                 vocab_size=1, freeze=False,
+                 **kwargs):
+        super(TransformerDecoder, self).__init__()
+
+        # build all (num_layers) layers
+        layers = []
+        for _ in range(num_layers):
+            layer = TransformerDecoderLayer(
+                hidden_size,
+                MultiHeadedAttention(num_heads, hidden_size),
+                MultiHeadedAttention(num_heads, hidden_size),
+                PositionwiseFeedForward(hidden_size, ff_size, dropout), dropout)
+            layers.append(layer)
+
+        self.layers = nn.ModuleList(layers)
+        self.norm = nn.LayerNorm(hidden_size)
+        self.pe = PositionalEncoding(hidden_size, dropout=dropout)
+
+        self._hidden_size = hidden_size
+        self._output_size = vocab_size
+
+        self.output_layer = nn.Linear(hidden_size, vocab_size, bias=False)
+
+        if freeze:
+            freeze_params(self)
+
+    def forward(self,
+                trg_embed: Tensor = None,
+                encoder_output: Tensor = None,
+                encoder_hidden: Tensor = None,
+                src_mask: Tensor = None,
+                unrol_steps: int = None,
+                hidden: Tensor = None,
+                trg_mask: Tensor = None,
+                **kwargs):
+        """
+        :param trg_embed: target embeddings
+        :param encoder_output: source representations
+        :param encoder_hidden: unused
+        :param src_mask:
+        :param unrol_steps: unused
+        :param hidden: unused
+        :param trg_mask:
+        :param kwargs:
+        :return:
+        """
+        assert trg_mask is not None, "trg_mask required for Transformer"
+
+        x = trg_embed
+        x = self.pe(x)  # add position encoding
+
+        trg_mask = trg_mask.unsqueeze(-2) & \
+            subsequent_mask(trg_embed.size(1)).type_as(trg_mask)
+
+        for layer in self.layers:
+            x = layer(x, encoder_output, src_mask, trg_mask)
+
+        x = self.norm(x)
+        output = self.output_layer(x)
+
+        return output, x, None, None
+
+    def __repr__(self):
+        return "%s(num_layers=%r, num_heads=%r)" % (
+            self.__class__.__name__, len(self.layers),
+            self.layers[0].self_attn.h)
+
+
+class TransformerDecoderLayer(nn.Module):
+    """
+    Transformer decoder layer.
+    Consists of self-attention, source-attention, and feed-forward.
+    """
+
+    def __init__(self, size, self_attn, src_attn, feed_forward, dropout):
+        super(TransformerDecoderLayer, self).__init__()
+        self.size = size
+        self.self_attn = self_attn
+        self.src_attn = src_attn
+        self.feed_forward = feed_forward
+        self.sublayer = clones(SublayerConnection(size, dropout), 3)
```

Comment:
hard-coded 3: make configurable?

Reply:
Can we make the `SublayerConnection` name more meaningful? Without knowing the Annotated Transformer it needs quite some digging to find out what's happening there.

Reply:
Nope ;-)

Reply:
Nope? No more meaning? :laughing:

Reply:
Was renamed to ResidualLayer.